### ISSUE TITLE
[Machine Learning] Gradient Descent

### ISSUE THREAD

**Issue by one-pieces** [one-pieces/practices#3]
>注：本文为台湾大学李宏毅教授的《机器学习》课程的 [Gradient Descent](https://www.bilibili.com/video/av10590361/?p=6) 一章的笔记。

# 一、梯度下降 Gradient Descent
机器学习的第三个步骤是，我们要找到一个最好的 function。这需要找到参数 θ 的最优解，即新的参数 θ* 要比之前的参数 θ 对结果更优。

<img width="740" alt="screen shot 2019-02-01 at 4 19 34 pm" src="https://user-images.githubusercontent.com/13073614/52111017-6de38a80-263d-11e9-9d2a-8e94eec3451a.png">

##### 勘误：argmax 应为 argmin。

定义新参数 θ* 与现参数 θ 的函数关系为 L， 称 L 为 Loss 函数。我们的目标是通过学习迭代，求出 Loss 函数的最优解。

## 1. 如何迭代
其实从上图可以看到，现参数 θ 加上学习率 η 乘上偏导数 ▽C，就得到 θ*。
<img width="738" alt="screen shot 2019-02-01 at 4 48 15 pm" src="https://user-images.githubusercontent.com/13073614/52112463-91a8cf80-2641-11e9-9ab8-f6af9ca78d70.png">
如此反复，最终达到最优值。
<img width="739" alt="screen shot 2019-02-01 at 5 01 18 pm" src="https://user-images.githubusercontent.com/13073614/52112986-0b8d8880-2643-11e9-974e-84f80352bf24.png">

### 2.1. 小心设置学习率 Learning Rate
如果学习率设置过小，则学习步长会很小，导致学习过程变长；如果过大，则可能导致无法达到最优点。

<img width="737" alt="screen shot 2019-02-01 at 5 13 22 pm" src="https://user-images.githubusercontent.com/13073614/52113585-b488b300-2644-11e9-8dd7-c56838f4c611.png">

### 2.2. 自动设置学习率
调学习率是一个比较麻烦的事情，那我们可不可以做一些规则去设置呢？
#### 2.2.1 每次迭代的学习率都不同
在刚开始时，起始点离目标点还比较远，所以学习步长可以大一点。当几次迭代后，离目标点较近，则步长需要小一点。比如，第 t 次的学习率 ηt = η / √(t+1)。
#### 2.2.2 每个参数的学习率都不同
比如 Adagrad。
<img width="739" alt="screen shot 2019-02-01 at 5 30 51 pm" src="https://user-images.githubusercontent.com/13073614/52114493-25c96580-2647-11e9-87fd-49f5f3ea90ca.png">

# 二、Adagrad
Adagrad 结合了上面说的每次迭代且每个参数的学习率都不同的两个规则。

<img width="739" alt="screen shot 2019-02-01 at 5 43 38 pm" src="https://user-images.githubusercontent.com/13073614/52115213-06333c80-2649-11e9-9c9e-bfccbc2b818b.png">

具体例子可看下图。

<img width="738" alt="screen shot 2019-02-01 at 5 50 44 pm" src="https://user-images.githubusercontent.com/13073614/52115593-ecdec000-2649-11e9-93f4-4488259c9775.png">

最终，Adagrad 的式子为

<img width="738" alt="screen shot 2019-02-01 at 5 53 22 pm" src="https://user-images.githubusercontent.com/13073614/52115727-48a94900-264a-11e9-9f8f-4709a9df28e3.png">

### 1.1 放大反差
一般的梯度下降，gradient 越大，学习步长就越大。但从上面的式子可以看出，Adagrad 的下图的“矛盾”。

<img width="736" alt="screen shot 2019-02-03 at 10 41 12 am" src="https://user-images.githubusercontent.com/13073614/52171904-40632200-27a0-11e9-8ca8-fc02889427b5.png">

如何解释这个“矛盾”呢？其实这个可以理解为当前 gradient 与之前所有 gradient 的“反差”。

<img width="741" alt="screen shot 2019-02-03 at 10 46 34 am" src="https://user-images.githubusercontent.com/13073614/52171952-00506f00-27a1-11e9-8e11-4411fd6bc5d4.png">

当前 gradient 比之前所有 gradient 之和大很多时，学习率乘以的系数就会比当前 gradient 大很多。反之当小很多时，其系数就比原本的 gradient 小很多。 

#### 1.2.1 一维变量
当只有一维变量时，起始点为 <a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;$x_0$" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\inline&space;$x_0$" title="$x_0$" /></a>，则最好的步长 <a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;$\left|&space;x_0&space;&plus;&space;\cfrac{b}{2a}&space;\right|$" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\inline&space;$\left|&space;x_0&space;&plus;&space;\cfrac{b}{2a}&space;\right|$" title="$\left| x_0 + \cfrac{b}{2a} \right|$" /></a>，即 <a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;$\cfrac{\left|&space;2ax_0&space;&plus;b&space;\right|}{2a}$" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\inline&space;$\cfrac{\left|&space;2ax_0&space;&plus;b&space;\right|}{2a}$" title="$\cfrac{\left| 2ax_0 +b \right|}{2a}$" /></a>

<img width="742" alt="screen shot 2019-02-03 at 11 12 40 am" src="https://user-images.githubusercontent.com/13073614/52172150-a6ea3f00-27a4-11e9-8736-473adf218662.png">

可以看出，最好步长跟一次导数 <a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;$\left|&space;2ax_0&space;&plus;b&space;\right|$" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\inline&space;$\left|&space;2ax_0&space;&plus;b&space;\right|$" title="$\left| 2ax_0 +b \right|$" /></a> 成正比。一次导数越大，则步长越大。

但这只是在只有一维变量的情况下成立，当有多个变量时就不成立了。

比如下图有两个变量，a 点的导数比 b 点的导数大，所以 a 点离最低点比较远。同样 c 点导数比 d 点大，也是离最低点比较远。但拿 a 点和 c 点作比较时，我们发现 c 点的导数比 a 点的大，然而 c 点比 a 点离最低点近。

<img width="740" alt="screen shot 2019-02-03 at 11 41 13 am" src="https://user-images.githubusercontent.com/13073614/52172401-60e3aa00-27a9-11e9-923a-5b029c5f7445.png">

#### 1.2.2 多个变量
最好的步长是 <a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;$\cfrac{\left|&space;2ax_0&space;&plus;b&space;\right|}{2a}$" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\inline&space;$\cfrac{\left|&space;2ax_0&space;&plus;b&space;\right|}{2a}$" title="$\cfrac{\left| 2ax_0 +b \right|}{2a}$" /></a>。当有多个变量时，我们应该还要考虑分母部分，即 <a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;$2a$" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\inline&space;$2a$" title="$2a$" /></a>。而这个 <a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;$2a$" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\inline&space;$2a$" title="$2a$" /></a> 其实就是 y 的二次导数。

<img width="869" alt="screen shot 2019-02-09 at 12 48 04 am" src="https://user-images.githubusercontent.com/13073614/52492536-635c5e80-2c04-11e9-919e-c08dbe1f26e5.png">

即最好步长应为 <a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;$\cfrac{\left|&space;\cfrac{\partial&space;y}{\partial&space;x}&space;\right|}{\cfrac{\partial^2&space;y}{\partial&space;x^2}}$" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\inline&space;$\cfrac{\left|&space;\cfrac{\partial&space;y}{\partial&space;x}&space;\right|}{\cfrac{\partial^2&space;y}{\partial&space;x^2}}$" title="$\cfrac{\left| \cfrac{\partial y}{\partial x} \right|}{\cfrac{\partial^2 y}{\partial x^2}}$" /></a>。

回到刚才的问题，我们可以看到，在 W1 的方向上的二次导数是比较小的，其曲线比较平滑。在 W2 的方向上二次导数比较大，所以曲线比较陡。

<img width="869" alt="screen shot 2019-02-09 at 12 57 20 am" src="https://user-images.githubusercontent.com/13073614/52493116-ae2aa600-2c05-11e9-9b7e-3b2ab9af4177.png">

#### 1.2.3 二次导数的近似
很多时候我们需要学习很多参数的值，如果每次都求二次导数的话，计算成本会很高。因此我们可以用之前已经计算好的多个一次导数的值近似的估算二次导数的值。

<img width="871" alt="screen shot 2019-02-09 at 1 03 49 am" src="https://user-images.githubusercontent.com/13073614/52493610-e4b4f080-2c06-11e9-83a1-c61e57f21db3.png">

# 三、随机梯度下降 Stochastic Gradient Descent
